zenoh-kotlin/src/commonMain/kotlin/io/zenoh/selector/Selector.kt

```diff
@@ -27,7 +28,25 @@ import io.zenoh.keyexpr.KeyExpr
  * @property keyExpr The [KeyExpr] of the selector.
  * @property parameters The parameters of the selector.
  */
-class Selector(val keyExpr: KeyExpr, val parameters: String = ""): AutoCloseable {
+class Selector(val keyExpr: KeyExpr, val parameters: String = "") : AutoCloseable {
+
+    val parametersMap = deserializeParametersToMap()
+
+    private fun deserializeParametersToMap(): Map<String, String?> {
+        val keyValuePairs = parameters.split('&').map { parameter ->
+            if (parameter.contains('=')) {
+                parameter.split('=').let { it[0] to URLDecoder.decode(it[1], Charsets.UTF_8) }
+            } else {
+                parameter to null
+            }
+        }
+        validateParameters(keyValuePairs)
+        return keyValuePairs.toMap()
+    }
+
+    private fun validateParameters(keyValuePairs: List<Pair<String, String?>>) = keyValuePairs
+        .groupBy({ it.first }, { it.second })
+        .forEach { (name, values) -> if (values.size > 1) throw Exception("Duplicated parameter $name detected") }
```

Comment:
A couple of comments:

* The functionality you are providing here is the one from [`pub fn parameters_stringmap(&'a self) -> ZResult<HashMap<String, String>>`](https://docs.rs/zenoh/latest/zenoh/selector/struct.Selector.html#method.parameters_stringmap) on Rust. 
The [parameters_map](https://docs.rs/zenoh/latest/zenoh/selector/struct.Selector.html#method.parameters_map) function is similar but providing generic arguments. We will implement it in the future.

* For the sake of API alignment, hese functions should be treated as functions and not as attributes

* Similar to Rust's Results (or to our custom implementation `ZResult`), Kotlin has its Result functionality which we use extensively across this library instead. Let's stick to that policy.

* A key without value shouldn't return null as value, but an empty string. See the RFC for the Selectors [https://github.com/eclipse-zenoh/roadmap/tree/main/rfcs/ALL/Selectors#the-parameters] where it's explained.

* I think we can slightly improve the code, to make it more 'kotlinish' and to break the execution earlier in case of key duplication.


Therefore I suggest the following changes:

```suggestion
    /** Extracts the selector [parameters]' name-value pairs into a map, returning an error in case of duplicated parameters. */
    fun parametersStringMap(): Result<Map<String, String>> = runCatching {
        val parametersMap = mutableMapOf<String, String>()
        parameters.split('&').forEach { parameter ->
            val pair = parameter.split('=')
            val key = pair[0]
            if (parametersMap.containsKey(key)) {
                throw IllegalArgumentException("Duplicated parameter `$key` detected.")
            }
            val value = pair.getOrNull(1)?.let { value -> URLDecoder.decode(value, Charsets.UTF_8) } ?: ""
            parametersMap[key] = value
        }
        parametersMap.toMap()
    }
```
WDYT?

Reply:
Hey, thank you for the feedback on this, inspired from your approach I wrote another similar implementation which instead of having a `mutableMap` + `forEach` makes use of the `fold` method. 
What do you think? If you find it harder to read or less idiomatic I can just use your version 👍🏻

Reply:
```kotlin
    /** Extracts the selector [parameters]' name-value pairs into a map, returning an error in case of duplicated parameters. */
    fun parametersStringMap(): Result<Map<String, String>> = runCatching {
        parameters.split('&').fold(mapOf()) { parametersMap, parameter ->
            val keyValuePair = parameter.split('=')
            val key = keyValuePair[0]
            if (parametersMap.containsKey(key)) {
                throw IllegalArgumentException("Duplicated parameter `$key` detected.")
            }
            val value = keyValuePair.getOrNull(1)?.let { URLDecoder.decode(it, Charsets.UTF_8) } ?: ""
            parametersMap + (key to value)
        }
    }
```